tensorflow/compiler/xla/service/gpu/gpu_fused_mha_runner.cc

```diff
@@ -0,0 +1,520 @@
+  const Shape &lhs_bmm1_shape = desc.lhs_bmm1_shape;
+  const Shape &rhs_bmm1_shape = desc.rhs_bmm1_shape;
+  const Shape &rhs_bmm2_shape = desc.rhs_bmm2_shape;
+  const Shape &intermediate_lhs_bmm2_shape = desc.intermediate_lhs_bmm2_shape;
+  const Shape &output_shape = desc.output_shape;
+
+  // Get DNN dtype from primtive types
+  TF_ASSIGN_OR_RETURN(
+      DataType lhs_bmm1_type,
+      GetDNNDataTypeFromPrimitiveType(lhs_bmm1_shape.element_type()));
+  TF_ASSIGN_OR_RETURN(
+      DataType rhs_bmm1_type,
+      GetDNNDataTypeFromPrimitiveType(rhs_bmm1_shape.element_type()));
+
+  TF_ASSIGN_OR_RETURN(
+      DataType rhs_bmm2_type,
+      GetDNNDataTypeFromPrimitiveType(rhs_bmm2_shape.element_type()));
+  TF_ASSIGN_OR_RETURN(DataType lhs_bmm2_type,
+                      GetDNNDataTypeFromPrimitiveType(
+                          intermediate_lhs_bmm2_shape.element_type()));
+  TF_ASSIGN_OR_RETURN(DataType output_type, GetDNNDataTypeFromPrimitiveType(
+                                                output_shape.element_type()));
+  GpufMHAConfig config;
+  config.input_type = lhs_bmm1_shape.element_type();
+  config.output_type = output_shape.element_type();
+
+  // Get MatmulTensorDescriptors for BMM1
+  config.lhs_bmm1 =
+      MatmulTensorDescriptor::For(lhs_bmm1_type, lhs_bmm1_shape.dimensions(),
+                                  desc.lhs_bmm1_shape.layout().minor_to_major(),
+                                  desc.bmm1_dnums.lhs_batch_dimensions(),
+                                  desc.bmm1_dnums.lhs_contracting_dimensions());
+  config.rhs_bmm1 =
+      MatmulTensorDescriptor::For(rhs_bmm1_type, rhs_bmm1_shape.dimensions(),
+                                  desc.rhs_bmm1_shape.layout().minor_to_major(),
+                                  desc.bmm1_dnums.rhs_batch_dimensions(),
+                                  desc.bmm1_dnums.rhs_contracting_dimensions());
+
+  // Get MatmulTensorDescriptors for BMM2
+  config.rhs_bmm2 =
+      MatmulTensorDescriptor::For(rhs_bmm2_type, rhs_bmm2_shape.dimensions(),
+                                  desc.rhs_bmm2_shape.layout().minor_to_major(),
+                                  desc.bmm2_dnums.rhs_batch_dimensions(),
+                                  desc.bmm2_dnums.rhs_contracting_dimensions());
+
+  config.intermediate_lhs_bmm2 = MatmulTensorDescriptor::For(
+      lhs_bmm2_type, intermediate_lhs_bmm2_shape.dimensions(),
+      desc.intermediate_lhs_bmm2_shape.layout().minor_to_major(),
+      desc.bmm2_dnums.lhs_batch_dimensions(),
+      desc.bmm2_dnums.lhs_contracting_dimensions());
+
+  config.output = TensorDescriptor::For(output_type, output_shape.dimensions(),
+                                        output_shape.layout().minor_to_major());
+
+  config.kind = desc.kind;
+  const CudnnfMHABackendConfig &backend_config = desc.backend_config;
+  config.algorithm = se::dnn::AlgorithmDesc(backend_config.algorithm());
+
+  auto check_and_assign_mask = [&]() -> Status {
```

Comment:
Nit: stateful procedures are a bit hard to read, maybe would be more readable if moved out into separate functions with switch inside?

Reply:
Are you referring to separate function that has the `switch (config.kind)` within and get rid of the lambdas somehow? I am not quite clear. Can u please elaborate a bit?

Reply:
Addressed in the commit https://github.com/tensorflow/tensorflow/pull/60233/commits/9297a101eda9423629edfdaee45dc1baf265305a